torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp

```diff
@@ -0,0 +1,115 @@
+#include <shared_mutex>
+
+#include <torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+namespace {
+
+class ProcessGlobalProfilerStateStackEntry;
+
+#if defined(__MACH__)
+// Compiler error: 'shared_timed_mutex' is unavailable: introduced in
+// macOS 10.12
+std::mutex
+#else
+std::shared_timed_mutex
+#endif
+    currentProcessGlobalProfilerStateStackEntryMutex;
+std::shared_ptr<ProcessGlobalProfilerStateStackEntry>
+    currentProcessGlobalProfilerStateStackEntryPtr = nullptr;
+
+class ProcessGlobalProfilerStateStackEntry {
+ public:
+  static std::shared_ptr<ProcessGlobalProfilerState> current() {
+#if defined(__MACH__)
+    std::unique_lock<std::mutex>
+#else
+    std::shared_lock<std::shared_timed_mutex>
+#endif
+        rlock(currentProcessGlobalProfilerStateStackEntryMutex);
+
+    if (!currentProcessGlobalProfilerStateStackEntryPtr) {
+      return nullptr;
+    }
+    return currentProcessGlobalProfilerStateStackEntryPtr
+        ->profilerProcessGlobalStatePtr_;
+  }
+
+  static void push(std::shared_ptr<ProcessGlobalProfilerState>
+                       profilerProcessGlobalStatePtr) {
+#if defined(__MACH__)
+    std::unique_lock<std::mutex>
+#else
+    std::unique_lock<std::shared_timed_mutex>
+#endif
+        wlock(currentProcessGlobalProfilerStateStackEntryMutex);
+
+    auto previousStateStackEntryPtr =
+        currentProcessGlobalProfilerStateStackEntryPtr;
+    currentProcessGlobalProfilerStateStackEntryPtr =
+        std::make_shared<ProcessGlobalProfilerStateStackEntry>();
+    currentProcessGlobalProfilerStateStackEntryPtr
+        ->previousProcessGlobalProfilerStateStackEntryPtr_ =
+        previousStateStackEntryPtr;
+    currentProcessGlobalProfilerStateStackEntryPtr
+        ->profilerProcessGlobalStatePtr_ =
```

Comment:
Is there any standard way to make good acronyms for variable names? If ProcessGlobalProfiler is widely used in our code base, shall we just call it `Pgp` for abbreviation? :)

Reply:
Let me put them into namespace to remove common prefix.